PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py

```diff
@@ -787,40 +814,54 @@ def getMETUncertainties(self, process, metType, metModName, electronCollection,
                                            veto = jetCollection,
                                            useDeltaRforFootprint = cms.bool(False)
                                            )
-            if self._parameters["Puppi"].value:
+            if self.getvalue("Puppi"):
               pfCandsNoJets.useDeltaRforFootprint = True
             addToProcessAndTask("pfCandsNoJets"+postfix, pfCandsNoJets, process, task)
             metUncSequence += getattr(process, "pfCandsNoJets"+postfix)
 
             #electron projection ==
             pfCandsNoJetsNoEle = cms.EDProducer("CandPtrProjector", 
                                                 src = cms.InputTag("pfCandsNoJets"+postfix),
-                                                veto = electronCollection
+                                                veto = electronCollection,
+                                                useDeltaRforFootprint = cms.bool(False),
                                                 )
+            if not self.getvalue("onMiniAOD"):
+              pfCandsNoJetsNoEle.useDeltaRforFootprint = True
+              pfCandsNoJetsNoEle.veto = "pfeGammaToCandidate:electrons"
```

Comment:
This line:
fCandsNoJetsNoEle.veto = "pfeGammaToCandidate:electrons"
sets process.pfeGammaToCandidate.electrons = "pfeGammaToCandidate:electrons"
a very weird bug in the config system?

Reply:
I changed it back to
pfCandsNoJetsNoEle.veto = cms.InputTag("pfeGammaToCandidate","electrons")
But maybe some followup on the input tag assignment is needed...

Reply:
@makortel @Dr15Jones 
please check.
I was hoping that the right hand side ``` "pfeGammaToCandidate:electrons"``` will be equivalent to``` cms.InputTag("pfeGammaToCandidate","electrons")``` and will also allow to not specify the type for an existing config parameter.
Instead we get a modification to something elsewhere

Reply:
I did
```
>>> import FWCore.ParameterSet.Config as cms
>>> p = cms.PSet(a = cms.InputTag("foo"))
>>> print p
cms.PSet(
    a = cms.InputTag("foo")
)
>>> p.a = "bar"
>>> print p
cms.PSet(
    a = cms.InputTag("bar")
)
>>> p.a = "bar:fii"
>>> print p
cms.PSet(
    a = cms.InputTag("bar","fii")
)
```
and it worked just fine.

Reply:
Switching to using an EDProducer also works
```python
>>> p = cms.EDProducer("DummyProd", a=cms.InputTag("foo"))
>>> print p.a
cms.InputTag("foo")
>>> p.a = "bar"
>>> print p.a
cms.InputTag("bar")
>>> p.a = "bar:fii"
>>> print p.a
cms.InputTag("bar","fii")
```

Reply:
Are you certain the original type was a `cms.InputTag` and not a `cms.string`?

Reply:
Is the symptom that an `InputTag` is modified here, and some other `InputTag` elsewhere gets the same change?

Reply:
```python
>>> a = cms.PSet(a = cms.InputTag("foo:bar"))
>>> a
cms.PSet(
    a = cms.InputTag("foo","bar")
)
>>> b = cms.PSet(b = a.a)
>>> b
cms.PSet(
    b = cms.InputTag("foo","bar")
)
>>> a.a = "fii:xyzzy"
>>> b.b
cms.InputTag("fii","xyzzy")
```
vs.
```python
>>> a = cms.PSet(a = cms.InputTag("foo:bar"))
>>> a
cms.PSet(
    a = cms.InputTag("foo","bar")
)
>>> import copy
>>> b = cms.PSet(b = copy.copy(a.a))
>>> b
cms.PSet(
    b = cms.InputTag("foo","bar")
)
>>> a.a = "fii:xyzzy"
>>> b.b
cms.InputTag("foo","bar")
```

Reply:
IIUC the symptom is  ```modA.parA = "modB:parB"```
leads to  ```modB.parB = "modB:parB"```

Reply:
It is InputTag. I'm not able to reproduce the issue with a simple example. However, it can be reproduced by a printout on the line referenced here:

print(process.pfeGammaToCandidate.electrons)
pfCandsNoJetsNoEle.veto = "pfeGammaToCandidate:electrons"
print(process.pfeGammaToCandidate.electrons)

with output:

cms.InputTag("selectedPatElectrons")
cms.InputTag("pfeGammaToCandidate","electrons")

If I do:
print(process.pfeGammaToCandidate.electrons)
pfCandsNoJetsNoEle.veto = cms.InputTag("pfeGammaToCandidate","electrons")
print(process.pfeGammaToCandidate.electrons)

I get output:

cms.InputTag("selectedPatElectrons")
cms.InputTag("selectedPatElectrons")

Reply:
Where are `pfCandsNoJetsNoEle.veto` and `process.pfeGammaToCandidate.electrons` set for the first time?

Reply:
A simple example however, can't reproduce the issue:

from __future__ import print_function
import FWCore.ParameterSet.Config as cms
from PhysicsTools.PatAlgos.tools.helpers import getPatAlgosToolsTask, addToProcessAndTask
process = cms.Process("PAT")
from PhysicsTools.PatUtils.pfeGammaToCandidate_cfi import pfeGammaToCandidate
task = getPatAlgosToolsTask(process)
addToProcessAndTask("pfeGammaToCandidate", pfeGammaToCandidate.clone(electrons = cms.InputTag("selectedPatElectrons")), process, task)
pfCandsNoJetsNoEle = cms.EDProducer("CandPtrProjector", src = cms.InputTag("pfCandsNoJets"), veto = cms.InputTag("selectedPatElectrons"), print(process.pfeGammaToCandidate.electrons)
pfCandsNoJetsNoEle.veto = "pfeGammaToCandidate:electrons"
print(process.pfeGammaToCandidate.electrons)

Reply:
https://github.com/cms-sw/cmssw/blob/9d6e4c7e0179f2b9c74fcda653ee63fa9cb5318e/PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py#L371

https://github.com/cms-sw/cmssw/blob/9d6e4c7e0179f2b9c74fcda653ee63fa9cb5318e/PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py#L825

Reply:
instead of veto, is maybe electronCollection itself modified by the assignement?

Reply:
>  addToProcessAndTask("pfeGammaToCandidate", pfeGammaToCandidate.clone(electrons = cms.InputTag("selectedPatElectrons")), process, task)

can this be a problem somehow that the cloned source and the assigned to names are the same?

Reply:
Thanks @ahinzmann. Indeed both of them share the same `InputTag` object, and therefore if one is modified the new value propagates to *all* instances. Assigning `= cms.InputTag(...)` makes `pfCandsNoJetsNoEle.veto` to point to a *different* `InputTag` object.

Reply:
Basically it would be best to do
```cpp
import copy
def someFunction(process, someInputTag):
    ...
    process.someModule.parameter = copy.copy(someInputTag)
    process.someOtherModule.parameter = copy.copy(someInputTag)
```

Reply:
this sounds like we found a conflict between the idea to drop cms.Type specifications to avoid accidentally inserting a new parameter and python language itself.
What would be a good resolution?

Is there a way to enforce ```modA.parA = cms.SomeType(valA)``` not to insert a new parameter into modA if parA does not already exist?

Reply:
> this sounds like we found a conflict between the idea to drop cms.Type specifications to avoid accidentally inserting a new parameter and python language itself.
> What would be a good resolution?

Always copy/clone a `cms.X` when such an object is potentially assigned to multiple variables. (I think this has been the recommendation for a long time)

> Is there a way to enforce `modA.parA = cms.SomeType(valA)` not to insert a new parameter into modA if parA does not already exist?

Technically yes, but it would be a deliberate, and likely breaking, change to current behavior. For example, under that behavior it would not be possible to add new parameters with `modA.parA = cms.SomeType(valA)`.

Reply:
I think my previous question came from a confusion and this point of using cms.SomeType is a digression.
To clarify, 
IIUC the following 
modA.parA = xyz
modB.parB = xyz
modA.parA = "zzz"
```
would actually change modB.parB. This is fine and is a case where a copy is indeed needed.

In the earlier comment though I somehow ended up assuming that literals/values are the same objects, like in 
```
modA.parA = "xyz"
modB.parB = "xyz"
modA.parA = "zzz"
```
the last line would also affect the value of modB.parB

Reply:
> In the earlier comment though I somehow ended up assuming that literals/values are the same objects, like in
> 
> ```
> modA.parA = "xyz"
> modB.parB = "xyz"
> modA.parA = "zzz"
> ```
> 
> the last line would also affect the value of modB.parB

I agree it is confusing. The literals themselves are the same objects, but `parA` and `parB` are (well, may be) different because they are attributes of an `_Parameterizable` ("a PSet-like") object. Part of the confusion (at least for me) is that the "assignment operation" is a `__setattr__()` call on `modA`, not an "`operator=()` call on `modA.parA`.

E.g. in your example, if `parB` is created as a reference to `parA`, the last line *does* affect the value of `parB`
```python
>>> modA.parA = cms.string("xyz")
>>> modA.parB = modA.parA
>>> modA.parA = "zzz"
>>> modA.parB
cms.string('zzz')
```
but if it is (re)created as a new object, the last line *does not* affect the value of `parB`
```python
>>> modA.parA = cms.string("xyz")
>>> modA.parB = cms.string("xyz")
>>> modA.parA = "zzz"
>>> modA.parB
cms.string('xyz')
```